lib/perform_bulk.rb

```diff
@@ -0,0 +1,284 @@
+  #     end
+  #   end
+  #
+  # but after awhile, this job has high throughput, and we'd greatly benefit
+  # from bulk inserting such simple log rows.
+  #
+  # we can modify it to be a bulk job:
+  #
+  #   class AuditLogJob
+  #     include Sidekiq::Job
+  #     include PerformBulk::Job
+  #
+  #     def perform(*logs)
+  #       AuditLog.insert_all(logs)
+  #     end
+  #   end
+  #
+  # we can queue jobs normally, but during times of high throughput, they'll
+  # be batched up according to job class and executed in bulk:
+  #
+  #   25.times { AuditLogJob.perform_async(...) }
+  #
+  # the above will result in a single job execution, where the *logs job args
+  # are equal to the args of the 25 batched jobs.
+  module Job
+    def self.included(klass)
+      raise ArgumentError, "cannot be used outside of Sidekiq job (got #{klass.ancestors})" unless
+        klass < Sidekiq::Job
+
+      klass.extend ClassMethods
+
+      # NB(ezekg) swap the bulk job's queue over to our capsule's ingress queue in
+      #           case our job doesn't ever call .sidekiq_options
+      klass.sidekiq_options queue: QUEUE_WAITING
+    end
+
+    module ClassMethods
+      # FIXME(ezekg) there's gotta be a better way to do this?
+      def sidekiq_options(...)
+        opts = super(...)
+
+        # hijack bulk job to push new jobs to our capsule's ingress queue, and
+        # then we'll push to the expected egress queue after batching.
+        unless opts['queue'] == QUEUE_WAITING
+          queue_was  = opts.delete('queue') || QUEUE_DEFAULT
+          queue      = QUEUE_WAITING
+
+          Logger.instance.debug(trace: :job) { "hijacking #{name.inspect} from #{queue_was.inspect} to #{queue.inspect}" }
+
+          opts['queue_was'] = queue_was
+          opts['queue']     = queue
+        end
+
+        opts
+      end
+    end
+  end
+
+  class Logger
+    include Singleton
+
+    %i[debug info warn error fatal].each do |level|
+      define_method(level) do |*msgs, **tags, &block|
+        log(level, *msgs, **tags, &block)
+      end
+    end
+
+    private
+
+    def log(level, *msgs, **tags, &block)
+      Sidekiq::Context.with(trace: [LOG_PREFIX, tags.delete(:trace)].compact.join('.'), **tags) do
+        Sidekiq.logger.send(level, *msgs, &block)
+      end
+    end
+  end
+
+  module Logging
+    class Chain
+      def initialize(**tags) = @default_tags = tags.freeze
+
+      %i[debug info warn error fatal].each do |level|
+        define_method(level) do |*msgs, **tags, &block|
+          Logger.instance.send(level, *msgs, **default_tags.merge(tags), &block)
+        end
+      end
+
+      private
+
+      attr_reader :default_tags
+    end
+
+    def self.[](trace, **)
+      Module.new do
+        include Logging
+
+        define_method :logger do
+          @_logger ||= Logging::Chain.new(**, trace:)
+        end
+
+        private :logger
+      end
+    end
+
+    private
+
+    def logger = Logger.instance
+  end
+
+  class Processor
```

Comment:
Can we make this so that it shows batch size in Sidekiq logs? e.g. `PerformBulk::Processor[100]: start|done`

Reply:
https://github.com/sidekiq/sidekiq/blob/b647e3adfe28be3bdefd90aabc345eeb46202591/lib/sidekiq/job_logger.rb#L30